### ISSUE TITLE
Precompile binary distributions?

### ISSUE THREAD

**Issue by jheintz** [BrianPugh/python-template#38]
After looking in way too many google searches I ended up getting 'python setup.py bdist_wheel' to build a precompiled manylinux wheel by setting is_pure to False

    class BinaryDistribution(Distribution):
        def is_pure(self):
            return False

in https://github.com/GistLabs/cythonwheel/blob/c3998d2ca74a512eca61762c4193c7f92b80d3e9/setup.py

How do I instruct Cython and cibuildwheels to do the same thing? Thanks!

**Comment by BrianPugh** [comment]
I usually just let Github Actions handle all the wheel building (it'll happen automatically whenever you create a `vX.Y.Z` tag) and uploading to pypi. This includes building a manylinux wheel (see relevant part of [action workflow](https://github.com/BrianPugh/python-template/blob/9e217e7d27ee91c7da70ef7c55bd2445b47c685e/.github/workflows/build_wheels.yaml#L90)).

If you need to build a wheel locally, running `poetry build` will create a wheel reflecting your local virtual environment, like `dist/foobar-1.1.4.post23+bc5f3ef-cp311-cp311-macosx_13_0_arm64.whl` (My venv is using cpython 3.11 running on maxosx 13.0 using an M1 chip). The versioning is handled by [poetry-dynamic-versioning](https://github.com/mtkennerly/poetry-dynamic-versioning), so the plugin has to be installed (run `poetry self add "poetry-dynamic-versioning[plugin]"`)

**Comment by jheintz** [comment]
I must be doing something wrong... the .c code doesn't seem to be getting compiled in the cibuildwheels action. The source code ends up in the .whl files (both the .c and the .py).

In this build 
https://github.com/jheintz/librarytemplate/actions/runs/5564911019
there are several whl files that seem to be successfully built, but they contain the source code (and the cythonized .c code from the py files) but they don't contain compiled object libraries. This is the contents of one of the whl files that is produced:

`
john@mba2020 Downloads % find librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info/RECORD
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info/LICENSE
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info/WHEEL
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info/entry_points.txt
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate-0.0.0.post8+6f76c03.dist-info/METADATA
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/.DS_Store
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/_c_extension.cpython-311-x86_64-linux-gnu.so
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/mod1.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/__main__.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/__init__.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/__init__.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/mod1.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/mod2.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/__init__.html
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/__init__.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/__init__.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/main.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/main.html
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/cli/main.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/_c_extension.pyx
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/mod2.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/clibrarytemplate.pxd
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/py.typed
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/_c_src
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/_c_src/foo.c
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/_c_src/foo.h
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate/__main__.py
librarytemplate-0.0.0.post8+6f76c03-cp311-cp311-manylinux_2_17_x86_64.manylinux_2_5_x86_64.manylinux1_x86_64.manylinux2014_x86_64.whl/librarytemplate.libs`

**Comment by BrianPugh** [comment]
I inspected a few of the wheels in your artifacts, and they include compiled objects (e.g. `_c_extension.cpython-311-x86_64-linux-gnu.so`). Not sure why one of your builds is failing, but we can investigate that if we'd like.

**Comment by jheintz** [comment]
Of course you're right that there was a .so file in there... what I meant to say was that the *.py files aren't compiled into .so libraries. 

If manually delete the mod2.py and mod2.c files (and the lines from dist-info/RECORD) then the whl file will install into a Colab linux but will fail to load:
![Screenshot 2023-07-18 at 11 41 32 AM](https://github.com/BrianPugh/python-template/assets/72560/77bc629b-cbed-4172-92b3-6d8493836d0e)

**Comment by jheintz** [comment]
What I was expecting is that the *.py (and the *.c) files would all be missing from the binary whl distribution files and each would be compiled to a *.o file (or whatever for platform).

This is the contents of the `python setup.py bdist_wheel` output in https://github.com/GistLabs/cythonwheel/actions/runs/5588436817 project:
![Screenshot 2023-07-18 at 11 48 47 AM](https://github.com/BrianPugh/python-template/assets/72560/88f3ac7c-ee3c-4fbb-a80d-76514c3521e2)

**Comment by BrianPugh** [comment]
investigating

**Comment by BrianPugh** [comment]
I added a possible solution to your issues as a PR: https://github.com/jheintz/librarytemplate/pull/1

While coming up with that solution, I discovered a few related shortcomings in the template that I've already fixed or will fix soon. Unfortunately, I don't think I can directly add the boilerplate for cythonizing python files as the exact actions are more bespoke to each project. Hopefully this issue is a good starting point for others that run into similar issues.

Does this solve your problem?

**Comment by jheintz** [comment]
Woohoo! I might have gotten it working! At least I haven't found any problems so far and it loaded correctly into Google Colab so that is something! This is the latest on main and builds at least some of the artifacts :) https://github.com/jheintz/librarytemplate/actions/runs/5616580760 

<img width="1115" alt="Screenshot 2023-07-20 at 5 26 42 PM" src="https://github.com/BrianPugh/python-template/assets/72560/60efcfb1-104c-44b2-96ea-ff52f353e2f7">

The code you provided in PR was really useful. It did have the disadvantage that nested python modules didn't work because they all got moved and overwritten into the top module directory.

In https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html I saw a reference to Extension("*", ["*.pyd"]) that turned out really helpful. I _think_ that the "*" is a magic character for find all the files and do the right thing with them. Seems to be working for my purposes.
